### ISSUE TITLE
inconsistent result of ```tf.raw_ops.BiasAddGrad``` on CPU and GPU

### ISSUE THREAD

**Issue by usmonali4** [tensorflow/tensorflow#86378]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

tf 2.18

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.11

### CUDA/cuDNN version

12.5/9

### GPU model and memory

Tesla T4

### Current behavior?

```tf.raw_ops.BiasAddGrad``` produces inconsistent results between CPU and GPU

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

out_backprop = tf.constant([
    [
        [
            [[ 0.2207,  2.1094], [-0.3730, -1.0625], [ 1.7031,  0.7148]], 
            [[ 1.5078, -0.6719], [-0.6367,  0.5039], [-2.3281,  0.5078]]
        ],
        [
            [[-0.3574,  0.0461], [ 2.3750, -2.9688], [-0.5703, -2.0156]],
            [[ 0.8125,  1.7656], [-0.9570,  0.6250], [-0.6914, -0.4746]]
        ],
        [
            [[-0.3750, -0.7383], [ 0.3691,  0.4570], [ 1.1641,  0.2715]],
            [[-1.2969, -0.9844], [-0.4863,  1.0938], [-1.4297,  0.8086]]
        ]
    ],
    [
        [
            [[ 0.3730,  0.8477], [-0.3887,  1.2266], [ 0.0859, -0.5742]],
            [[-0.7383, -0.2432], [-0.7578, -0.8281], [-0.1660, -0.9336]]
        ],
        [
            [[ 1.4297,  0.6797], [-1.6172,  0.4941], [-0.3047, -0.3711]],
            [[-0.6250, -0.7617], [ 0.9453,  0.1064], [ 1.4062, -2.9531]]
        ],
        [
            [[-1.4297, -0.1387], [ 0.0625,  1.0469], [-0.1953,  1.6406]],
            [[-0.3047,  0.5117], [ 1.8125,  1.1797], [-0.8789, -0.4688]]
        ]
    ]
], dtype=tf.bfloat16)

with tf.device('CPU:0'):
  result_cpu = tf.raw_ops.BiasAddGrad(out_backprop=out_backprop, data_format="NCHW")
  print("BiasAddGrad Output on CPU:", result_cpu)

with tf.device('GPU:0'):
  result_gpu = tf.raw_ops.BiasAddGrad(out_backprop=out_backprop, data_format="NCHW")
  print("BiasAddGrad Output on GPU:", result_gpu)

max_abs_diff = tf.reduce_max(tf.abs(result_cpu - result_gpu)).numpy()

is_consistent = tf.experimental.numpy.allclose(tf.cast(result_cpu, tf.float32), tf.cast(result_gpu, tf.float32), rtol=1e-3,  atol=1e-2)

print("Max absolute difference:", max_abs_diff)
print("Consistency check (CPU vs GPU) with atol=1e-2 and rtol=1e-3:", is_consistent.numpy())
```

### Relevant log output

```shell
BiasAddGrad Output on CPU: tf.Tensor([0.09375 -3.96875 1.70312], shape=(3,), dtype=bfloat16)

BiasAddGrad Output on GPU: tf.Tensor([0.078125 -4 1.70312], shape=(3,), dtype=bfloat16)

Max absolute difference: 0.03125
Consistency check (CPU vs GPU) with atol=1e-2 and rtol=1e-3: False
```

**Comment by tilakrayal** [comment]
@usmonali4,
I think the issue is specific to GPU, I was able to reproduce the issue on colab with GPU runtime. But, when the inputs are changed to float64 precision, the results are as expected.
The reason could be due to the optimized way of calculating numbers on Nvidia which is different compared to others, there will be a small amount of precision errors.
The same behavior is not observed on Apple M1, using numpy or in CPU.
https://github.com/tensorflow/tensorflow/issues/58479

Thank you!

**Comment by usmonali4** [comment]
@tilakrayal,
similar reasoning as in this comment [#86256/comment](https://github.com/tensorflow/tensorflow/issues/86256#issuecomment-2636423339), but the acceptable tolerance thresholds have been set to atol=1e-2 and rtol=1e-3 considering that the dtypes of results are bfloat16 (16‑bit floating‐point types). Glad to hear your thoughts on this; thanks for your time.

**Comment by tilakrayal** [comment]
@usmonali4,
Looks like the other issues which were created are similar to #86256 . Could you please close the other issues which were raised which helps to track the issue in a better way. Thank you!

**Comment by usmonali4** [comment]
@tilakrayal,
if you take for example this report and #86256, I believe that the core issue is different since they are targeting different api's with different types of tensors

**Comment by jobs-git** [comment]
Reproduced here:

```bash
TensorFlow Version: 2.16.2
Python Version: 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
Python Implementation: CPython
CUDA Version (TensorFlow): 12.0
cuDNN Version (TensorFlow): 8
GPU Name (PyTorch): NVIDIA GeForce RTX 4070 SUPER
CUDA Version (PyTorch): 12.0
cuDNN Version (PyTorch): 8907
Driver Version: 550.144.03
Ubuntu Version: PRETTY_NAME="Ubuntu 24.04.1 LTS"
Model name:                           AMD Ryzen 5 7600 6-Core Processor


BiasAddGrad Output on CPU: tf.Tensor([0.0625 -3.96875 1.70312], shape=(3,), dtype=bfloat16)
BiasAddGrad Output on GPU: tf.Tensor([0.078125 -4 1.70312], shape=(3,), dtype=bfloat16)
Max absolute difference: 0.03125
Consistency check (CPU vs GPU) with atol=1e-2 and rtol=1e-3: False

```

**Comment by avmohansai** [comment]
Hi @usmonali4  ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

From the example shared, this appears to be related to expected numerical precision differences between CPU and GPU implementations of `tf.raw_ops.BiasAddGrad` when using the bfloat16 datatype. Since bfloat16 has limited mantissa precision, small rounding differences can occur across different hardware backends and reduction kernels, particularly for accumulation operations such as gradient reductions.

In the reproducer, the observed differences are relatively small (0.03125) and are likely caused by backend-specific floating point accumulation and rounding behavior rather than a functional correctness issue. If stricter numerical consistency is required across devices, it may help to use higher precision datatypes such as `float32` or relax the comparison tolerances used in validation checks.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
